Packages/com.unity.inputsystem/InputSystem/Plugins/Switch/SwitchProControllerHID.cs

```diff
@@ -113,12 +129,352 @@ namespace UnityEngine.InputSystem.Switch
     /// <summary>
     /// A Nintendo Switch Pro controller connected to a desktop mac/windows PC using the HID interface.
     /// </summary>
-    /// <remarks>
-    /// This controller currently only works when connected via Bluetooth but not when connected over USB.
-    /// </remarks>
     [InputControlLayout(stateType = typeof(SwitchProControllerHIDInputState), displayName = "Switch Pro Controller")]
-    public class SwitchProControllerHID : Gamepad
+    public class SwitchProControllerHID : Gamepad, IInputStateCallbackReceiver, IEventPreProcessor
     {
+        protected override void OnAdded()
+        {
+            base.OnAdded();
+            HandshakeRestart();
+        }
+
+        private static readonly SwitchMagicOutputReport.CommandIdType[] s_HandshakeSequence = new[]
+        {
+            SwitchMagicOutputReport.CommandIdType.Status,
+            SwitchMagicOutputReport.CommandIdType.Handshake,
+            SwitchMagicOutputReport.CommandIdType.Highspeed,
+            SwitchMagicOutputReport.CommandIdType.Handshake,
+            SwitchMagicOutputReport.CommandIdType.ForceUSB
+            ////TODO: Should we add a step to revert back to simple interface?
+            //// Because currently full reports don't work in old input system.
+        };
+
+        private int m_HandshakeStepIndex;
+        private double m_HandshakeTimer;
+
+        private void HandshakeRestart()
+        {
+            // Delay first command issue until some time into the future
+            m_HandshakeStepIndex = -1;
+            m_HandshakeTimer = InputRuntime.s_Instance.currentTime;
+        }
+
+        private void HandshakeTick()
+        {
+            const double handshakeRestartTimeout = 2.0;
+            const double handshakeNextStepTimeout = 0.1;
+
+            var currentTime = InputRuntime.s_Instance.currentTime;
+
+            // There were no events for last few seconds, restart handshake
+            if (currentTime >= m_LastUpdateTimeInternal + handshakeRestartTimeout &&
+                currentTime >= m_HandshakeTimer + handshakeRestartTimeout)
+                m_HandshakeStepIndex = 0;
+            // If handshake is complete, ignore the tick.
+            else if (m_HandshakeStepIndex + 1 >= s_HandshakeSequence.Length)
+                return;
+            // If we timeout, proceed to next step after some time is elapsed.
+            else if (currentTime > m_HandshakeTimer + handshakeNextStepTimeout)
+                m_HandshakeStepIndex++;
+            // If we haven't timed out on handshake step, skip the tick.
+            else
+                return;
+
+            m_HandshakeTimer = currentTime;
+
+            var command = s_HandshakeSequence[m_HandshakeStepIndex];
+
+            // Native backend rejects one of the commands based on size of descriptor.
+            // So just report both at a same time.
+            ////TODO: fix this.
+            var commandBt = SwitchMagicOutputHIDBluetooth.Create(command);
```

Comment:
should we move this to jira as a ticket to fix at some point? does it need fixing? what's the outcome of leaving as is vs fixing?

Reply:
ISX-956